### ISSUE TITLE
Undocumented File Format Structures (as of 0.99)

### ISSUE THREAD

**Issue by starknebula** [ephtracy/voxel-model#19]
There exists undocumented data members inside the VOX file format. Proceeding the XYZI data, we can see tags labelled:

1. nTRN
2. nGRP
3. nSHP
4. LAYR

These appear to be non-standard when compared to the other Chunk types as they don't seem to be 4-byte aligned and are may be terminated with 0xFFFFFFFF.

Furthermore the old description for materials MATT appears to have been replaces by a new descriptor MATL which uses strings to identify material properties rather than pure binary.

**Comment by daishihmr** [comment]
Thanks a lot!
I will also investigate them.

**Comment by starknebula** [comment]
If there's anything I can do to help let me know.

**Comment by sh-dave** [comment]
Please update the docs further, `LAYR` still isn't documented, also `rOBJ` seems to be missing.

**Comment by scayze** [comment]
bump. LAYR and rOBJ is still not documented. would be very appreciated.

**Comment by Snkz** [comment]
theres also DOCT

**Comment by aiekick** [comment]
Hello all,

like you, i tried to understand the new file format. i thinck im successfull.

the weird thing about this new format, is the format number (150) is the same in each voxel files i opened ....

i used hexamonkey (http://hexamonkey.com), who is a fabulous binary file analyser,
 for tried to write the grammar fo the two vox file.

so i found that the byte offset you seen, is due to the STRING TYPE in each DICT.

and i think the LAYER Chunck is :

![notepad _2018-05-29_18-48-24](https://user-images.githubusercontent.com/1434736/40673050-e884acaa-6370-11e8-909c-7bddc2b5528f.png)

i seen that the unknown var is always -1, so i think its a reserved int on 4 bytes, like other reserved items.

 so it display that :

![hexamonkey_2018-05-29_18-47-14](https://user-images.githubusercontent.com/1434736/40672997-b8fa9c6a-6370-11e8-84fc-41c3197ae118.png)

here the grammar: 

![notepad _2018-05-29_18-42-01](https://user-images.githubusercontent.com/1434736/40672773-160b7c72-6370-11e8-8cc2-af6b566a38a2.png)
![notepad _2018-05-29_18-42-17](https://user-images.githubusercontent.com/1434736/40672775-16470bb6-6370-11e8-8161-e51e8740b7c4.png)
![notepad _2018-05-29_18-42-24](https://user-images.githubusercontent.com/1434736/40672774-162a15b0-6370-11e8-8259-bf954d1f380c.png)

and here the result : 

![hexamonkey_2018-05-29_18-37-00](https://user-images.githubusercontent.com/1434736/40672632-a8b575ec-636f-11e8-9e62-daa878808c89.png)

marvelous no ? :)

**Comment by scayze** [comment]
Thank you! This is amazing!

**Comment by aiekick** [comment]
i not seen the rOBj in my vox files. do you have sample vox file with this rOBJ ? for search the structure

**Comment by FlayaN** [comment]
@DanielGilbert I created a pullrequest with the rOBJ chunk https://github.com/aiekick/MagicaVoxel_File_Formats/pull/1

**Comment by aiekick** [comment]
i have approved the request. :)

**Comment by aiekick** [comment]
i added in my repo a basis vox writer with world mode support :) it can be improved i think but for the moment it take the job. my goal was just to export voxel block in new world mode from my Soft "SdfMesher" : https://twitter.com/hashtag/sdfmesher?src=hash

**Comment by alexhunsley** [comment]
Hi all, found this issue because I'm working on a numpy-vox format lib and I also noticed stuff in the saved vox files not in the format. 
Just adding a +1 to the request for updated file format docs.

**Comment by alexhunsley** [comment]
Or in the very least, updating the model version written to files (apparently it's still at 150), so we can at least know if we're trying to open a file with a model format we don't yet know how to parse.

**Comment by ephtracy** [comment]
the reason why the version is still 150 is to make the file readable in older versions (the old version only accepts files with version 150).

**Comment by ephtracy** [comment]
the rendering setting are not open yet because they are changing frequently. But you can still read it since it is just in the DICT format. I will update the format of the Layer chunks though.

**Comment by CubixSystem** [comment]
> I will update the format of the Layer chunks though.

@ephtracy Could you please provide an update on .vox format documentation?

**Comment by princec** [comment]
+1 ... would rather like to know what rOBJ does

**Comment by princec** [comment]
(I'm guessing it's lighting, background, skybox and camera information, and I really would like to use that so I can import scenes directly into my game engine)

**Comment by mgerhardy** [comment]
What is IMAP about? A chunk I've found in the PrinceOfPersia scene: https://www.patreon.com/posts/prince-of-persia-26600425

**Comment by hugodecasta** [comment]
About the **IMAP** and **NOTE** chunk:

**IMAP**:
Represents the **palette "Index" Map**
size = 256
256 x int(1 byte)

The palette of a VOX file contains **256 colors**, all identified by their **index** and containing a **specific color (r,g,b,a)**.

The RGBA chunk **contains the color information** of the palette **but not the exact indexes** (let's assume for the moment that the order of RGBA is random).

To overcome this problem and find the exact palette index of each color, the IMAP chunk stores the **RGBA/palette index mapping**.

* **Example**:
RGBA = `[{ r:148, g: 63, b:125, a:255 }, { r:32, g: 54, b:53, a:255 }, { r:0, g: 25, b:1, a:255 }, ...]`
IMAP = `[3, 5,  41, 63, 84, 201, 17, ...]`
In this voxel file, the color `{ r:0, g: 25, b:1, a:255 }` (index `2` in RGBA) 's true palette index is `41`.

By the way ... `voxel color = RGBA[voxel.index - 1]`

**NOTE**:
Contains all **Color type Names**
 - int32 : number of names (would be 32)
 - for each name: 
       - [STRING](https://github.com/ephtracy/voxel-model/blob/eed2ea10cfc03890b80787cb19ebfafbadb2670d/MagicaVoxel-file-format-vox-extension.txt#L7)

**Comment by TomWor** [comment]
@hugodecasta Thanks for the hint regarding IMAP, saved my day!
I had a model where in MV I would see index 1 as a certain color, but while importing and reading the file, it insisted the index is 242. The voxel had that index before, but I would later change it in MV, so the index (and RGBA order) must have stayed the same. Very confusing.

**Comment by hugodecasta** [comment]
> @hugodecasta Thanks for the hint regarding IMAP, saved my day!
> I had a model where in MV I would see index 1 as a certain color, but while importing and reading the file, it insisted the index is 242. The voxel had that index before, but I would later change it in MV, so the index (and RGBA order) must have stayed the same. Very confusing.

Yes, the RGBA data saving order is still a mystery to me too (maybe it has to do with the inner working of MV: moving stuff around for other purposes. Who knows).

**Comment by TomWor** [comment]
@hugodecasta Yes. Also: When IMAP chunk is not present (needs to be checked when importing), the index of the RGBA chunk and voxel color index match, like one would expect.

**Comment by hugodecasta** [comment]
> @hugodecasta Yes. Also: When IMAP chunk is not present (needs to be checked when importing), the index of the RGBA chunk and voxel color index match, like one would expect.

Until an official behavior documentation hard to find out why such different behaviors
